src/kernel/tty.zig

```diff
@@ -0,0 +1,2523 @@
+        p.* = blank;
+    }
+}
+
+///
+/// When the text/terminal gets to the bottom of the screen, then move all line up by the amount
+/// that are below the bottom of the screen. Usually moves up by one line.
+///
+/// Global changes:
+///     pages
+///     video_buffer
+///     row
+///
+/// Arguments:
+///
+/// Return:
+///
+fn scroll() void {
+    // Added the condition in the if from pagesMoveRowsUp as don't need to move all rows
+    if (row >= vga.HEIGHT and (row - vga.HEIGHT + 1) <= ROW_TOTAL) {
+        var rows_to_move: u16 = row - vga.HEIGHT + 1;
+
+        // Move rows up pages by temp, will usually be one.
+        pagesMoveRowsUp(rows_to_move);
+
+        // Move all rows up by rows_to_move
+        var i: u16 = 0;
+        while (i < (ROW_TOTAL - rows_to_move) * vga.WIDTH) : (i += 1) {
+            video_buffer[START_OF_DISPLAYABLE_REGION + i] = video_buffer[(rows_to_move * vga.WIDTH) + START_OF_DISPLAYABLE_REGION + i];
+        }
+
+        // Set the last rows to blanks
+        i = 0;
+        while (i < vga.WIDTH * rows_to_move) : (i += 1) {
+            video_buffer[(vga.HEIGHT - rows_to_move) * vga.WIDTH + i] = blank;
+        }
+
+        row = vga.HEIGHT - 1;
+    }
+}
+
+///
+/// Print a character without updating the cursor. For speed when printing a string as only need to
+/// update the cursor once. This will also print the special characters: \n, \r, \t and \b. (\b is
+/// not a valid character so use \x08 which is the hex value).
+///
+/// Global changes:
+///     pages
+///     video_buffer
+///     row
+///     column
+///
+/// Arguments:
+///     IN char: u8 - The character to print.
+///
+/// Return:
+///
+fn _putChar(char: u8) void {
+    switch (char) {
+        '\n' => {
+            column = 0;
+            row += 1;
+            scroll();
+        },
+        '\t' => {
+            column += 4;
+            if (column >= vga.WIDTH) {
+                column -= @truncate(u8, vga.WIDTH);
+                row += 1;
+                scroll();
+            }
+        },
+        '\r' => {
+            column = 0;
+        },
+        // \b
+        '\x08' => {
+            if (column == 0) {
+                if (row != 0) {
+                    column = vga.WIDTH - 1;
+                    row -= 1;
+                }
+            } else {
+                column -=  1;
+            }
+        },
+        else => {
+            putEntryAt(char, column, row);
+            column += 1;
+            if (column == vga.WIDTH) {
+                column = 0;
+                row += 1;
+                scroll();
+            }
+        },
+    }
+}
+
+///
+/// Print a character without updating the cursor. For speed when printing a string as only need to
+/// update the cursor once. This will also print the special characters: \n, \r, \t and \b
+///
+/// Global changes:
+///     video_buffer
+///
+/// Arguments:
+///     IN char: u8 - The character to print.
+///
+/// Return:
+///
+fn printLogo() void {
```

Comment:
I would say this isn't specific to the tty and could probably be moved into kmain.

Reply:
From chat, is staying in tty